### ISSUE TITLE
with 100 machines, can tiger beetle do 100M tps ?

### ISSUE THREAD

**Issue by zeroexcuses** [tigerbeetle/tigerbeetle#1143]
Does tigerbeetle scale horizontally? If we throw 100 machines at it, can we get 100 M tps:?

99.99% of transactions will only involve 2-3 accounts; some may involve 4. None will involve 5+.

EDIT: Consider a sim-city like simulation with 1B AI agents, each with their own inventory. I'm wondering if tigerbeetle could be used to track the inventory for such a system. Only read/write, no query needed.

**Comment by jorangreef** [comment]
Hey @zeroexcuses, thanks for the question!

> 99.99% of transactions will only involve 2-3 accounts; some may involve 4. None will involve 5+.

Yes, and we've seen this also in some of the payment switches we looked at.

TigerBeetle's performance is orthogonal to the cardinality of accounts. In fact, it can go slightly faster if you only have say 2 accounts, because these are then hot in L1 cache.

Whereas, horizontally-sharded systems would struggle to compete with this if both accounts are on different shards. They would always be paying the locking cost across shards before they can execute the transaction.

> Does tigerbeetle scale horizontally? If we throw 100 machines at it, can we get 100 M tps:?

By way of background, there's a great paper by Frank McSherry on this: https://www.usenix.org/system/files/conference/hotos15/hotos15-paper-mcsherry.pdf

TigerBeetle was designed to keep all data "core local", to move the code to the data, rather than the data to the code, so that the CPU does not need to wait on the latency of a network roundtrip. The reason is that, while network bandwidth is increasing (true), the speed of light in fiber is nevertheless pretty much fixed, so network roundtrip latencies are not increasing, and the cost of a network roundtrip is always going to be an eternity to the CPU.

The other factor to consider in this discussion, is that the double-entry workload is by nature highly contentious.

What I mean by this is that if you have 100m customers, but only 2-3 bank accounts, then, while you could horizontally shard over 100m customers, nevertheless each time you debit a customer, you still need to credit any of 2-3 bank accounts, so those 2-3 shards are going to serialize the performance of all other shards. Except, now you're also paying for network latency between shards.

The mistake that systems can make is to "go horizontal too soon". For example, to partition all keys across shards, from the moment they enter the system. Whereas, TigerBeetle is designed to process extremely hot data, and then you can start to think of horizontal sharding out to cold storage systems via storage tiering once the data cools.

We call this technique "diagonal scaling", being the combination of vertical scaling (optimal for the CPU) for hot data, with disaggregation of storage and compute (horizontal scaling) for cold data.

That said, you can always deploy multiple TB clusters and shard across them using TB's native two-phase transfers. (In fact, it would be fun to try an experiment to see whether horizontally-sharded TB in this way can keep up with a single TB cluster!). We would rather give you the best possible latency before that scale, than make your transactions always touch shards and pay the latency cost of network RTTs, especially if you only have 2-3 accounts.

I go into this also in our recent talk with DuckDB Labs: https://www.youtube.com/watch?v=LHsjviZJ7PQ

**Comment by zeroexcuses** [comment]
1. Thank you for the detailed reply.
2. I think I incorrectly stated my problem (100% my fault), let me try to restate:

Suppose we try to build SimCityEarth: 10B.

1. We have 10B AI agents
2. Each agent has an "account"
3. Account is more than just bank balance, it also contains things like hit points, stamina points, items, etc ...
4. A transaction generally involves 2-3 accounts (I trade you gold in exchange for a diamond sword).
5. There can be up to 1B tps, but low contention. (I.e. if A is trading with B this second, they are unlikely to also be trading with C or D).

Is this problem a good match for TigerBeetle ? A common problem in game servers is item duplication or items vanishing during server crashes (or other edge cases). I am wondering if tigerbeetle is performant enough to just throw all virtual-agent-inventory at it.

Thanks!

**Comment by jorangreef** [comment]
@zeroexcuses I think I may have missed your edit, clarifying the contention issue.

If there's low contention, then to hit 1 billion TPS, you could explicitly do two-phase transfers across TigerBeetle clusters to get the throughput you need, linearly.

Just a heads up that we are tuning our new storage engine, which also does secondary indexing of all data (in addition to what the old engine used to do), to get it also up to 1m TPS. At present, it's close to 500k TPS with some optimizations about to come in.

Please join our Slack too. I'd be happy to chat live with you about this. Using TigerBeetle for game inventory is a really cool problem that we're excited about.

**Comment by slowfranklin** [comment]
Hm. I'm confused. Isn't VR a single leader protocol? Isn't the single leader a bottleneck serializing all operation? I was thrilled when I learned about ACCORD, a leaderless protocol, coming to Cassandra soon. Can you share your thoughts on this? Thanks!

**Comment by jorangreef** [comment]
> Isn't the single leader a bottleneck serializing all operation?

1. Not if the workload is itself serialized on hot accounts, which is true for most OLTP transactional workloads (e.g. debit a million customers, credit a single bank account).

2. Then, the TigerBeetle design is what we see as the fastest way to process transactions, since all hot accounts are kept core-local to a single CPU core.

3. Whereas another design would end up blocked on the network, and this decision would end up twice as expensive every two years (Moore's law is not yet dead, cf. the M3 Max with ~100 billion transistors) since network latencies (as opposed to bandwidth) remain constant (limited by the speed of light in fiber) whereas CPU transistors double roughly every two years.

4. TigerBeetle was purpose built for the OLTP transactional workload. For example, our VSR is also optimized for the OLTP workload and is not a generic consensus protocol. A single VSR primary is able to process thousands of transactions in a single RTT, and is generally not the bottleneck.

5. Finally, more important than scale alone, is the ability to hit massive scale—but with an efficient unit of scale. For example, because TB's unit of scale is as efficient as we could make it, a single cluster of 6 TB replicas might replace a horizontally scaled setup of 100 machines, where the unit of scale often gets less attention (no static memory allocation, no zero copy techniques etc.). Again, if you do need multiple units of scale (beyond 1m TPS) then you can always glue them together with explicit two-phase transfers across TB clusters, and that would probably be cheaper or more performant as well.

To summarize, you can trust TB to handle some extremely pathological workloads (that are surprisingly actually quite common in OLTP) with the same predictable performance, whereas utilizing multiple leaders would likely rapidly degrade (and not scale as efficiently).
